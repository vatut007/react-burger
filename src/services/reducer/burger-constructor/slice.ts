import { PayloadAction, createSlice, nanoid } from "@reduxjs/toolkit";
import {
  IngredientConstructorList,
  type Ingredient
} from "../../../types/ingredient";

export const burgerConstructorSlice = createSlice({
  name: "burgerConstructor",
  reducerPath: "burgerConstructor",
  initialState: {
    bun: null as Ingredient | null,
    ingredients: [] as IngredientConstructorList,
    ingredientCount: {} as Record<string,number>
  },
  reducers: {
    addIngredient: {
      reducer(
        state,
        action: PayloadAction<{ cartItemId: string; ingredient: Ingredient }>,
      ) {
        if (action.payload.ingredient.type === "bun") {
          state.bun = action.payload.ingredient;
        } else {
          state.ingredients.push({
            ...action.payload.ingredient,
            cart_item_id: action.payload.cartItemId,
          });
          if (action.payload.ingredient._id in state.ingredientCount){
            state.ingredientCount[action.payload.ingredient._id]++
          } else {
            state.ingredientCount[action.payload.ingredient._id]=1
          }
        }
      },
      prepare({ ingredient }: { ingredient: Ingredient }) {
        const cartItemId = nanoid();
        return { payload: { cartItemId, ingredient } };
      },
    },
    removeIngredient(state, action: PayloadAction<{ cartItemId: string, ingeredient_id: string }>) {
      state.ingredients = state.ingredients.filter(
        (item) => item.cart_item_id !== action.payload.cartItemId,
      );
      state.ingredientCount[action.payload.ingeredient_id]--
    },
    moveIngredient(
      state,
      action: PayloadAction<{ oldIndex: string; newIndex: number }>,
    ) {
      const index = state.ingredients.findIndex(
        (item) => item.cart_item_id === action.payload.oldIndex,
      );
      const [ingredient] = state.ingredients.splice(index, 1);
      state.ingredients.splice(action.payload.newIndex, 0, ingredient);
    },
  },
});
