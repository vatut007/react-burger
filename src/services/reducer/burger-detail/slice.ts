import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Ingredient } from "../../../types/ingredient";

export const ingredientDetailSlice = createSlice({
    name: 'ingedientDetail',
    reducerPath: 'ingedientDetail',
    initialState: {ingredient:null as Ingredient | null},
    reducers:{
        selectIngredient(state, action: PayloadAction<{ ingredient: Ingredient }>) {
            state.ingredient = action.payload.ingredient
        },
    }
})