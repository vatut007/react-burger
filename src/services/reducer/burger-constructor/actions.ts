import { burgerConstructorSlice } from "./slice";

export const { addIngredient, removeIngredient, moveIngredient } =
  burgerConstructorSlice.actions;
