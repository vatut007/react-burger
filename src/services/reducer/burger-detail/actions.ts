import { ingredientDetailSlice } from "./slice";

export const { selectIngredient } = ingredientDetailSlice.actions;
