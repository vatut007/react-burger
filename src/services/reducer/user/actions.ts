import { userSlice } from "./slice";

export const { addAccessToken, addUser, addRefreshToken, clearUser } =
  userSlice.actions;
