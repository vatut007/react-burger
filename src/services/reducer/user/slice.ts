import { PayloadAction, createSlice } from "@reduxjs/toolkit";
import { User } from "../../../types/user";

export const userSlice = createSlice({
  name: "user",
  reducerPath: "user",
  initialState: {
    user: null as User | null,
    accessToken: null as string | null,
    refreshToken: null as string | null,
  },
  reducers: {
    addUser(state, action: PayloadAction<User>) {
      state.user = {
        email: action.payload.email,
        name: action.payload.name,
      };
    },
    addAccessToken(state, action: PayloadAction<{ accessToken: string }>) {
      state.accessToken = action.payload.accessToken;
    },
    addRefreshToken(state, action: PayloadAction<{ refreshToken: string }>) {
      state.refreshToken = action.payload.refreshToken;
    },
    clearUser(state) {
      (state.accessToken = null),
        (state.refreshToken = null),
        (state.user = null);
    },
  },
});
