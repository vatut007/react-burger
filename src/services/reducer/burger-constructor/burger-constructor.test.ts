import { afterEach, beforeEach, expect, test, vi } from "vitest";
import type { Ingredient } from "../../../types/ingredient";
import { addIngredient, moveIngredient, removeIngredient } from "./actions";
import { burgerConstructorSlice } from "./slice";

const { nanoid } = vi.hoisted(() => ({ nanoid: vi.fn() }));

vi.mock("@reduxjs/toolkit", async (importOriginal) => {
  return {
    ...(await importOriginal<typeof import("@reduxjs/toolkit")>()),
    nanoid,
  };
});

const { reducer, getInitialState } = burgerConstructorSlice;

beforeEach(() => {
  nanoid
    .mockReturnValueOnce("random-string-first")
    .mockReturnValueOnce("random-string-second");
});

afterEach(() => {
  vi.restoreAllMocks();
});

const getTestBun = (): Ingredient => ({
  _id: "643d69a5c3f7b9001cfa093c",
  name: "Краторная булка N-200i",
  type: "bun",
  proteins: 80,
  fat: 24,
  carbohydrates: 53,
  calories: 420,
  price: 1255,
  image: "https://code.s3.yandex.net/react/code/bun-02.png",
  image_mobile: "https://code.s3.yandex.net/react/code/bun-02-mobile.png",
  image_large: "https://code.s3.yandex.net/react/code/bun-02-large.png",
  __v: 0,
});

const getTestIngredient = (): Ingredient => ({
  _id: "643d69a5c3f7b9001cfa0941",
  name: "Биокотлета из марсианской Магнолии",
  type: "main",
  proteins: 420,
  fat: 142,
  carbohydrates: 242,
  calories: 4242,
  price: 424,
  image: "https://code.s3.yandex.net/react/code/meat-01.png",
  image_mobile: "https://code.s3.yandex.net/react/code/meat-01-mobile.png",
  image_large: "https://code.s3.yandex.net/react/code/meat-01-large.png",
  __v: 0,
});

const getStateOneIngredient = () =>
  reducer(
    getInitialState(),
    addIngredient({ ingredient: getTestIngredient() }),
  );

const getStateTwoIngredients = () =>
  reducer(
    getStateOneIngredient(),
    addIngredient({ ingredient: getTestIngredient() }),
  );

test("add bun", () => {
  expect(
    reducer(getInitialState(), addIngredient({ ingredient: getTestBun() })),
  ).toMatchSnapshot();
});

test("add ingredient", () => {
  expect(
    reducer(
      getInitialState(),
      addIngredient({ ingredient: getTestIngredient() }),
    ),
  ).toMatchSnapshot();
});

test("add second ingredient", () => {
  expect(
    reducer(
      getStateOneIngredient(),
      addIngredient({ ingredient: getTestIngredient() }),
    ),
  ).toMatchSnapshot();
});

test("remove ingredient", () => {
  expect(
    reducer(
      getStateOneIngredient(),
      removeIngredient({
        cartItemId: "random-string-first",
        ingeredient_id: getTestIngredient()._id,
      }),
    ),
  ).toMatchSnapshot();
});

test("move ingredient", () => {
  expect(
    reducer(
      getStateTwoIngredients(),
      moveIngredient({ oldIndex: "random-string-first", newIndex: 1 }),
    ),
  ).toMatchSnapshot();
});
