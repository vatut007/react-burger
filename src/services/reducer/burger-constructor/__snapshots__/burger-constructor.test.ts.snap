// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`add bun 1`] = `
{
  "bun": {
    "__v": 0,
    "_id": "643d69a5c3f7b9001cfa093c",
    "calories": 420,
    "carbohydrates": 53,
    "fat": 24,
    "image": "https://code.s3.yandex.net/react/code/bun-02.png",
    "image_large": "https://code.s3.yandex.net/react/code/bun-02-large.png",
    "image_mobile": "https://code.s3.yandex.net/react/code/bun-02-mobile.png",
    "name": "Краторная булка N-200i",
    "price": 1255,
    "proteins": 80,
    "type": "bun",
  },
  "ingredientCount": {},
  "ingredients": [],
}
`;

exports[`add ingredient 1`] = `
{
  "bun": null,
  "ingredientCount": {
    "643d69a5c3f7b9001cfa0941": 1,
  },
  "ingredients": [
    {
      "__v": 0,
      "_id": "643d69a5c3f7b9001cfa0941",
      "calories": 4242,
      "carbohydrates": 242,
      "cart_item_id": "random-string-first",
      "fat": 142,
      "image": "https://code.s3.yandex.net/react/code/meat-01.png",
      "image_large": "https://code.s3.yandex.net/react/code/meat-01-large.png",
      "image_mobile": "https://code.s3.yandex.net/react/code/meat-01-mobile.png",
      "name": "Биокотлета из марсианской Магнолии",
      "price": 424,
      "proteins": 420,
      "type": "main",
    },
  ],
}
`;

exports[`add second ingredient 1`] = `
{
  "bun": null,
  "ingredientCount": {
    "643d69a5c3f7b9001cfa0941": 2,
  },
  "ingredients": [
    {
      "__v": 0,
      "_id": "643d69a5c3f7b9001cfa0941",
      "calories": 4242,
      "carbohydrates": 242,
      "cart_item_id": "random-string-first",
      "fat": 142,
      "image": "https://code.s3.yandex.net/react/code/meat-01.png",
      "image_large": "https://code.s3.yandex.net/react/code/meat-01-large.png",
      "image_mobile": "https://code.s3.yandex.net/react/code/meat-01-mobile.png",
      "name": "Биокотлета из марсианской Магнолии",
      "price": 424,
      "proteins": 420,
      "type": "main",
    },
    {
      "__v": 0,
      "_id": "643d69a5c3f7b9001cfa0941",
      "calories": 4242,
      "carbohydrates": 242,
      "cart_item_id": "random-string-second",
      "fat": 142,
      "image": "https://code.s3.yandex.net/react/code/meat-01.png",
      "image_large": "https://code.s3.yandex.net/react/code/meat-01-large.png",
      "image_mobile": "https://code.s3.yandex.net/react/code/meat-01-mobile.png",
      "name": "Биокотлета из марсианской Магнолии",
      "price": 424,
      "proteins": 420,
      "type": "main",
    },
  ],
}
`;

exports[`move ingredient 1`] = `
{
  "bun": null,
  "ingredientCount": {
    "643d69a5c3f7b9001cfa0941": 2,
  },
  "ingredients": [
    {
      "__v": 0,
      "_id": "643d69a5c3f7b9001cfa0941",
      "calories": 4242,
      "carbohydrates": 242,
      "cart_item_id": "random-string-second",
      "fat": 142,
      "image": "https://code.s3.yandex.net/react/code/meat-01.png",
      "image_large": "https://code.s3.yandex.net/react/code/meat-01-large.png",
      "image_mobile": "https://code.s3.yandex.net/react/code/meat-01-mobile.png",
      "name": "Биокотлета из марсианской Магнолии",
      "price": 424,
      "proteins": 420,
      "type": "main",
    },
    {
      "__v": 0,
      "_id": "643d69a5c3f7b9001cfa0941",
      "calories": 4242,
      "carbohydrates": 242,
      "cart_item_id": "random-string-first",
      "fat": 142,
      "image": "https://code.s3.yandex.net/react/code/meat-01.png",
      "image_large": "https://code.s3.yandex.net/react/code/meat-01-large.png",
      "image_mobile": "https://code.s3.yandex.net/react/code/meat-01-mobile.png",
      "name": "Биокотлета из марсианской Магнолии",
      "price": 424,
      "proteins": 420,
      "type": "main",
    },
  ],
}
`;

exports[`remove ingredient 1`] = `
{
  "bun": null,
  "ingredientCount": {
    "643d69a5c3f7b9001cfa0941": 0,
  },
  "ingredients": [],
}
`;
